### PR TITLE
Add support for drawing to multiple render targets

### LINKED ISSUE #6628: Support for drawing to multiple rendertargets
**Is your feature request related to a problem? Please describe (REQUIRED):**
Several graphical effects (e.g Many-light pipelines / deferred rendering, ssao, temporal antialiasing) operate on several components of a frame such as position, normals and albedo color. These are usually generated by drawing the scene once into several render targets that can be used in other parts of the rendering pipeline. MRT support is available from OpenGL 3 / ES3 / WebGL2 (and baseline vulkan), but is currently not exposed in any way.

**Describe the solution you'd like (REQUIRED):**
When creating a rendertarget in a render script, we could pass in extra params for the additional targets we wish to allocate:

Current API:
```
local color_params = { ... }
local depth_params = { ... }

self.my_render_target = render.render_target({
  [render.BUFFER_COLOR_BIT] = color_params,
  [render.BUFFER_DEPTH_BIT] = depth_params })`
```

**API idea 1:**

```
self.my_render_target = render.render_target({
	[render.BUFFER_COLOR_BIT_0] = color_params,
	[render.BUFFER_COLOR_BIT_1] = color_params,
	[render.BUFFER_COLOR_BIT_2] = color_params_with_float_type,
	[render.BUFFER_COLOR_BIT_3] = color_params_with_other_filtering,
	[render.BUFFER_DEPTH_BIT] = depth_params
})
```

**API idea 2:**

```
local color_params = {
	{...}, -- properties for buffer 1
	{...}, -- properties for buffer 2
	{...}, -- properties for buffer 3
	etc
}

self.my_render_target = render.render_target({
	[render.BUFFER_COLOR_BIT] = color_params,
	[render.BUFFER_DEPTH_BIT] = depth_params
})
```

**Additional context (OPTIONAL):**

Open questions:
OpenGL es2 / webgl 1 does support this functionality, but only through certain extensions which might not be universally supported. Right now we don't have a way to detect what extensions or features that are available outside the vanilla API, so for those graphics contexts, this would just not work. Do we want to have some way of detecting certain non-standard "features" in the engine somehow? If a certain platform doesn't support MRTs, a user might want to turn off certain effects, or do a workaround. Maybe something like render.has_support(render.MULTI_RENDERTARGET) or render.enable(render.MULTI_RENDERTARGET) that
a developer can use to decide how the render scripts should be configured.

### PR DESCRIPTION
**Pull request by Jhonnyg**
Many graphical effects (e.g Many-light pipelines / deferred rendering, ssao, temporal antialiasing) operate on several components of a frame such as position, normals and albedo color. These are usually generated by drawing the scene once into several render targets that can be used in other parts of the rendering pipeline. 

MRT is available from OpenGL 3 / ES3 / WebGL2 (and baseline vulkan) and support is now available in the Defold render.* API when creating a new render target:

```
function init(self)
    -- render target buffer parameters
    local color_params_rgba = { format = render.FORMAT_RGBA,
                                width = render.get_window_width(),
                                height = render.get_window_height(),
                                min_filter = render.FILTER_LINEAR,
                                mag_filter = render.FILTER_LINEAR,
                                u_wrap = render.WRAP_CLAMP_TO_EDGE,
                                v_wrap = render.WRAP_CLAMP_TO_EDGE }
    local color_params_float = { format = render.FORMAT_RG32F,
                           width = render.get_window_width(),
                           height = render.get_window_height(),
                           min_filter = render.FILTER_LINEAR,
                           mag_filter = render.FILTER_LINEAR,
                           u_wrap = render.WRAP_CLAMP_TO_EDGE,
                           v_wrap = render.WRAP_CLAMP_TO_EDGE }

    -- Create a render target with three color attachments
    -- Note: No depth buffer is attached here
    self.my_render_target = render.render_target({
           [render.BUFFER_COLOR0_BIT] = color_params_rgba,
           [render.BUFFER_COLOR1_BIT] = color_params_rgba,
           [render.BUFFER_COLOR2_BIT] = color_params_float, })
end

function update(self, dt)
    -- enable target so all drawing is done to it
    render.enable_render_target(self.my_render_target)
    -- draw a predicate to the render target
    render.draw(self.my_pred)
end
```

Fixes #6628

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/ShaderProgramBuilderTest.java [hunk 1]
```diff
@@ -200,11 +200,11 @@ public void testGlslDirectives() throws Exception {
                     "#extension GL_OES_standard_derivatives : enable\n" +
                     "precision mediump float;\n" +
                     "\n" +
-                    "out vec4 _DMENGINE_GENERATED_gl_FragColor;\n" +
+                    "out vec4 _DMENGINE_GENERATED_gl_FragColor_0;\n" +
```
Comment (Jhonnyg) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/ShaderProgramBuilderTest.java:203]: The GLSL transformation now supports multiple outputs, so the default will always be postfixed with _0.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderUtil.java [hunk 2]
```diff
@@ -246,12 +247,27 @@ public static Result transform(String input, ShaderType shaderType, String targe
...
-                input = input.replaceAll("\\b" + glFragColorKeyword + "\\b", glFragColorRep);
+            {
+                input = input.replaceAll("\\b" + glFragColorKeyword + "\\b", glFragColorRep + "_0");
+            }
+
             if (output_glFragData)
-                input = input.replaceAll("\\b" + glFragDataKeyword + "\\b", glFragColorRep);
+            {
+                input = input.replaceAll("\\b" + glFragDataKeyword + "\\[(\\d+)\\]", glFragColorRep + "_$1");
+            }
```
Comment (Jhonnyg) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderUtil.java:271]: Previously the gl_FragData keyword would be replaced with the same color output as gl_FragColor, but that doesn't work for multiple outputs. Instead we generate specific outputs for each assignment of gl_FragData[x] into gl_FragData_x variables.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderUtil.java [hunk 3]
```diff
@@ -333,8 +349,10 @@ else if (line.contains("precision")) {
                     if(floatPrecisionIndex < 0 && targetProfile.equals("es")) {
                         output.add(patchLineIndex++, floatPrecisionAttrRep);
                     }
-                    // insert fragcolor out attr
-                    output.add(patchLineIndex++, glFragColorAttrRep);
+                    for (int i = 0; i < maxColorOutputs; i++) {
+                         // insert fragcolor out attr
+                        output.add(patchLineIndex++, String.format(glFragColorAttrRep, "_" + i));   
+                    }
```
Comment (Jhonnyg) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderUtil.java:355]: There will always be at least one output

File: engine/graphics/src/graphics.h [hunk 4]
```diff
@@ -68,11 +68,17 @@ namespace dmGraphics
...
-        BUFFER_TYPE_COLOR_BIT   = 0x1,
-        BUFFER_TYPE_DEPTH_BIT   = 0x2,
-        BUFFER_TYPE_STENCIL_BIT = 0x4,
+        BUFFER_TYPE_COLOR_BIT   = 0x01,
+        BUFFER_TYPE_DEPTH_BIT   = 0x02,
+        BUFFER_TYPE_STENCIL_BIT = 0x04,
+        BUFFER_TYPE_COLOR0_BIT  = 0x01, // Equal to BUFFER_TYPE_COLOR_BIT
+        BUFFER_TYPE_COLOR1_BIT  = 0x08,
+        BUFFER_TYPE_COLOR2_BIT  = 0x20,
+        BUFFER_TYPE_COLOR3_BIT  = 0x40,
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/graphics/src/graphics.h:84]: These are internal values, and you should be able to do refactoring here.
What's important is the Lua script bindings.
They should use the Lua constants (which we define).
So in short, for the Lua script you should be able to do:

(pseudo code)
```
"BUFFER_TYPE_COLOR_BIT" = dmGraphics::BUFFER_TYPE_COLOR0_BIT;
"BUFFER_TYPE_COLOR0_BIT" = dmGraphics::BUFFER_TYPE_COLOR0_BIT;
```

File: engine/graphics/src/test/test_graphics.cpp [hunk 5]
```diff
@@ -588,7 +588,7 @@ TEST_F(dmGraphicsTest, TestGetRTAttachment)
         params[i].m_Width = WIDTH;
         params[i].m_Height = HEIGHT;
     }
-    assert(dmGraphics::MAX_BUFFER_TYPE_COUNT == 3);
+    assert(dmGraphics::MAX_BUFFER_TYPE_COUNT == 6);
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/graphics/src/test/test_graphics.cpp:633]: Perhaps a good code comment?

File: engine/graphics/src/test/wscript [hunk 6]
```diff
@@ -12,7 +12,7 @@ def build(bld):
                          uselib_local = 'graphics_null graphics_transcoder_null',
                          target = name)
 
-    print "MAWE", platform_supports_feature(bld.env.PLATFORM, 'vulkan', {}), bld.env.PLATFORM 
+    print "MAWE", platform_supports_feature(bld.env.PLATFORM, 'vulkan', {}), bld.env.PLATFORM
```
Comment (JCash) [inline engine/graphics/src/test/wscript:15]: it's a debug print, it shouldn't be there :)

File: engine/render/src/render/render_script.cpp [hunk 7]
```diff
@@ -2645,6 +2717,10 @@ namespace dmRender
         lua_setfield(L, -2, "BUFFER_"#name);
 
         REGISTER_BUFFER_CONSTANT(COLOR_BIT);
+        REGISTER_BUFFER_CONSTANT(COLOR0_BIT);
```
Comment (JCash) [inline engine/render/src/render/render_script.cpp:2720]: I think the dmGraphics::xxCOLOR_BIT should be removed, and instead be set here to COLOR0_BIT.

**Review by JCash — DISMISSED** [review]
Looks good, just a small enum change I think (if you agree)

File: engine/graphics/src/test/test_graphics.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jhonnyg) [reply engine/graphics/src/test/test_graphics.cpp:633]: 👍

File: engine/graphics/src/test/wscript (hunk 6 above)
Comment (Jhonnyg) [reply engine/graphics/src/test/wscript:15]: True :) I can remove it

File: engine/graphics/src/graphics.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jhonnyg) [reply engine/graphics/src/graphics.h:84]: True! That would make things easier yes, I mistakenly thought it was a value that was in the "wild" so to speak. I'll fix it

**Review by JCash — DISMISSED** [review]

File: engine/graphics/src/opengl/graphics_opengl.cpp [hunk 8]
```diff
@@ -2073,6 +2090,8 @@ static uintptr_t GetExtProcAddress(const char* name, const char* extension_name,
                 // attach the texture to FBO color attachment point
                 glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0 + i, GL_TEXTURE_2D, rt->m_ColorBufferTexture[i]->m_Texture, 0);
                 CHECK_GL_ERROR;
+
+                // printf("Creating RT at %d with %d \n", GL_COLOR_ATTACHMENT0 + i, rt->m_ColorBufferTexture[i]->m_Texture);
```
Comment (JCash) [inline engine/graphics/src/opengl/graphics_opengl.cpp:2099]: Might be worth having a "DEBUG" define at the top of the file, for easy enabling?

**Review by JCash — DISMISSED** [review]
lgtm

**Pull request merged**